### PR TITLE
module/prometheus: add grouping option

### PR DESCRIPTION
**Pull request by ilyam8**
Part of netdata/product#1308

This PR adds custom grouping feature for `Gauge` and `Counter` types metrics via `group` configuration option.

By default we create a chart per every metric name, `group` option allows to split those chart into multiple charts.

```yaml
jobs:
  - name: node_exporter_local
    url: http://127.0.0.1:9100/metrics
    group:
      - selector: <PATTERN>
        by_label: <a space separated list of labels names>
      - selector: <PATTERN>
        by_label: <a space separated list of labels names>
```

---

Example.

- metrics

```cmd
example_device_cur_state{name="0",type="Fan"} 0
example_device_cur_state{name="1",type="Fan"} 0
example_device_cur_state{name="10",type="Processor"} 0
example_device_cur_state{name="11",type="intel_powerclamp"} -1
example_device_cur_state{name="2",type="Fan"} 0
example_device_cur_state{name="3",type="Fan"} 0
example_device_cur_state{name="4",type="Fan"} 0
example_device_cur_state{name="5",type="Processor"} 0
example_device_cur_state{name="6",type="Processor"} 0
example_device_cur_state{name="7",type="Processor"} 0
example_device_cur_state{name="8",type="Processor"} 0
example_device_cur_state{name="9",type="Processor"} 0
```

- configuration

```yaml
jobs:
  - name: node_exporter_local
    url: http://127.0.0.1:9100/metrics
    group:
      - selector: example_device_cur_state
        by_label: type
```

- result

```cmd
# chart 1
example_device_cur_state{name="0",type="Fan"} 0
example_device_cur_state{name="1",type="Fan"} 0
example_device_cur_state{name="2",type="Fan"} 0
example_device_cur_state{name="3",type="Fan"} 0
example_device_cur_state{name="4",type="Fan"} 0

# chart 2
example_device_cur_state{name="11",type="intel_powerclamp"} -1

# chart 3
example_device_cur_state{name="10",type="Processor"} 0
example_device_cur_state{name="5",type="Processor"} 0
example_device_cur_state{name="6",type="Processor"} 0
example_device_cur_state{name="7",type="Processor"} 0
example_device_cur_state{name="8",type="Processor"} 0
example_device_cur_state{name="9",type="Processor"} 0
```

### CODE REVIEW

File: pkg/prometheus/selector/logical.go [hunk 1]
```diff
@@ -36,7 +41,7 @@ func Or(lhs, rhs Selector, others ...Selector) Selector {
 	return Or(m, others[0], others[1:]...)
 }
 
-// Not returns a matcher which opposites the sub-matcher's result
+// Not returns a selector which opposites the sub-selector's result
```
Comment (prologic) [inline pkg/prometheus/selector/logical.go:44]:
```suggestion
// Not returns a selector which returns the negation of the sub-selector's result
```

File: pkg/prometheus/selector/logical.go [hunk 2]
```diff
@@ -18,7 +18,12 @@ func (s negSelector) Matches(lbs labels.Labels) bool { return !s.s.Matches(lbs)
 func (s andSelector) Matches(lbs labels.Labels) bool { return s.lhs.Matches(lbs) && s.rhs.Matches(lbs) }
 func (s orSelector) Matches(lbs labels.Labels) bool  { return s.lhs.Matches(lbs) || s.rhs.Matches(lbs) }
 
-// And returns a matcher which returns true only if all of it's sub-matcher return true
+// TRUE returns a matcher which always returns true
+func TRUE() Selector {
```
Comment (prologic) [inline pkg/prometheus/selector/logical.go:22]: To be consistent with naming, `True`?

File: modules/prometheus/group.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+func contains(value string, in []string) bool {
+	switch len(in) {
+	case 0:
+		return false
+	case 1:
+		return value == in[0]
+	default:
+		return value == in[0] || contains(value, in[1:])
+	}
+}
```
Comment (prologic) [inline modules/prometheus/group.go:149]: Very nice use of tail-recursion here 😀

File: modules/prometheus/group.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+	"github.com/netdata/go.d.plugin/pkg/prometheus"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type grouping interface {
+	chartID(pm prometheus.Metric) string
+	dimID(pm prometheus.Metric) string
+	dimName(pm prometheus.Metric) string
+}
```
Comment (prologic) [inline modules/prometheus/group.go:15]: I don't really like this interface name or its descriptors. It feels very "un"-Go like 🤣

**Review by prologic — APPROVED** [review]

File: modules/prometheus/group.go (hunk 4 above)
Comment (ilyam8) [reply modules/prometheus/group.go:15]: i dont like it either 😆

**Pull request merged**